aya/src/maps/xdp/cpu_map.rs

```diff
@@ -0,0 +1,151 @@
+//! An array of available CPUs.
+
+use std::os::fd::AsRawFd;
+
+use aya_obj::generated::{bpf_cpumap_val, bpf_cpumap_val__bindgen_ty_1};
+
+use crate::{
+    maps::{check_bounds, check_kv_size, IterableMap, MapData, MapError},
+    sys::{bpf_map_lookup_elem, bpf_map_update_elem},
+    Pod,
+};
+
+/// An array of available CPUs.
+///
+/// XDP programs can use this map to redirect packets to a target
+/// CPU for processing.
+///
+/// # Minimum kernel version
+///
+/// The minimum kernel version required to use this feature is 4.15.
+///
+/// # Examples
+/// ```no_run
+/// # let elf_bytes = &[];
+/// use aya::maps::xdp::CpuMap;
+///
+/// let mut bpf = aya::BpfLoader::new()
+///     .set_max_entries("CPUS", aya::util::nr_cpus().unwrap() as u32)
+///     .load(elf_bytes)
+///     .unwrap();
+/// let mut cpumap = CpuMap::try_from(bpf.map_mut("CPUS").unwrap())?;
+/// let flags = 0;
+/// let queue_size = 2048;
+/// for i in 0u32..8u32 {
+///     cpumap.set(i, queue_size, None::<i32>, flags);
+/// }
+///
+/// # Ok::<(), aya::BpfError>(())
+/// ```
+#[doc(alias = "BPF_MAP_TYPE_CPUMAP")]
+pub struct CpuMap<T> {
+    inner: T,
+}
+
+impl<T: AsRef<MapData>> CpuMap<T> {
+    pub(crate) fn new(map: T) -> Result<CpuMap<T>, MapError> {
+        let data = map.as_ref();
+        check_kv_size::<u32, bpf_cpumap_val>(data)?;
+
+        let _fd = data.fd_or_err()?;
+
+        Ok(CpuMap { inner: map })
+    }
+
+    /// Returns the number of elements in the array.
+    ///
+    /// This corresponds to the value of `bpf_map_def::max_entries` on the eBPF side.
+    pub fn len(&self) -> u32 {
+        self.inner.as_ref().obj.max_entries()
+    }
+
+    /// Returns the value stored at the given index.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`MapError::OutOfBounds`] if `index` is out of bounds, [`MapError::SyscallError`]
+    /// if `bpf_map_lookup_elem` fails.
+    pub fn get(&self, index: u32, flags: u64) -> Result<CpuMapValue, MapError> {
+        let data = self.inner.as_ref();
+        check_bounds(data, index)?;
+        let fd = data.fd_or_err()?;
+
+        let value = bpf_map_lookup_elem(fd, &index, flags).map_err(|(_, io_error)| {
+            MapError::SyscallError {
+                call: "bpf_map_lookup_elem".to_owned(),
+                io_error,
+            }
+        })?;
+        let value: bpf_cpumap_val = value.ok_or(MapError::KeyNotFound)?;
+        Ok(value.into())
+    }
+
+    /// An iterator over the elements of the map. The iterator item type is `Result<u32,
+    /// MapError>`.
+    pub fn iter(&self) -> impl Iterator<Item = Result<CpuMapValue, MapError>> + '_ {
+        (0..self.len()).map(move |i| self.get(i, 0))
+    }
+}
+
+impl<T: AsMut<MapData>> CpuMap<T> {
+    /// Sets the value of the element at the given index.
```

Comment:
This needs better argument names (cpu_index and queue_size?), and needs to
explain how they work. What is `program`? More docs pls

Reply:
Indeed. I added quite a bit more in d79d8b4, which also specifies the constraints on what xdp programs can be loaded.